File: 7/blocks.cpp
#include <stack>
#include <vector>

int solution(vector<int> &H) {
    int blocks = 0;
    stack<int> stack;

    for(vector<int>::iterator it = H.begin(); it!=H.end(); it++) {
        while(!stack.empty() && stack.top() > *it)
            stack.pop();
        if (!stack.empty() && stack.top() == *it)
            ;
        else {
            blocks++;
            stack.push(*it);
        }
    }
    return blocks;
}

File: 7/blocks.py

def solution(H):
    stack = []
    blocks = 0

    for height in H:
        while len(stack) and stack[-1] > height:
            stack.pop()
        if len(stack) and stack[-1] == height:
            pass
        else:
            blocks += 1
            stack.append(height)
    return blocks